<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CU25_Modelo de gestión de Lista de Espera Quirúrgica
</font></center>


<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 07. Obtener un fichero con los datos sanitarios relevantes para las listas de espera.

* Los datos buenos habría que pedirlos al SNS. Mientras tanto, este archivo los simula en función de los datos existentes.
* En este notebook se guarda también un fichero de datos completo







<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Encoding

Con la siguiente expresión se evitan problemas con el encoding al ejecutar el notebook. Es posible que deba ser eliminada o adaptada a la máquina en la que se ejecute el código.

In [1]:
Sys.setlocale(category = "LC_ALL", locale = "es_ES.UTF-8")

[1] "es_ES.UTF-8/es_ES.UTF-8/es_ES.UTF-8/C/es_ES.UTF-8/C"

### Packages to use

_ELIMINAR O AÑADIR LO QUE TOQUE. COPIAR VERSIONES AL FINAL Y QUITAR CÓDIGO DE VERSIONES_

- {tcltk} para selección interactiva de archivos locales
- {sf} para trabajar con georeferenciación
- {readr} para leer y escribir archivos csv
- {dplyr} para explorar datos
- {stringr} para manipulación de cadenas de caracteres
- {tidyr} para organización de datos


In [2]:
library(readr)
library(dplyr)
library(tidyr)

p <- c("tcltk", "readr", "dplyr", "tidyr")



Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




### Paths

In [3]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)

1. Listas de espera para simular una capacidad acorde

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [4]:
# file_data_01 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [5]:
iFile_01 <- "CU_25_05_05_02_lista_espera.csv"
file_data_01 <- paste0(iPath, iFile_01)

if(file.exists(file_data_01)){
    cat("Se leerán datos del archivo: ", file_data_01)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leerán datos del archivo:  Data/Input/CU_25_05_05_02_lista_espera.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [6]:
data_01 <- read_csv(file_data_01)


Rows: 55680 Columns: 7
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): Hospital, Especialidad
dbl (5): total_pacientes, media_tiempo_dias, ano, semana, CODCNH

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [7]:
data_01 |> glimpse()

Rows: 55,680
Columns: 7
$ Hospital          <chr> "HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA A…
$ Especialidad      <chr> "Angiología y Cirugía Vascular", "Angiología y Cirug…
$ total_pacientes   <dbl> 439, 437, 429, 419, 418, NA, 408, 399, 396, 393, 378…
$ media_tiempo_dias <dbl> 56.29, 55.13, 53.99, 56.29, 55.82, NA, 57.45, 57.61,…
$ ano               <dbl> 2022, 2022, 2022, 2022, 2022, 2022, 2022, 2022, 2022…
$ semana            <dbl> 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, …
$ CODCNH            <dbl> 280148, 280148, 280148, 280148, 280148, 280148, 2801…


Muestra de datos:

In [8]:
data_01 |> slice_head(n = 5)

Hospital                                               
1 HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA ADELA
2 HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA ADELA
3 HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA ADELA
4 HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA ADELA
5 HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA ADELA
  Especialidad                  total_pacientes media_tiempo_dias ano  semana
1 Angiología y Cirugía Vascular 439             56.29             2022 44    
2 Angiología y Cirugía Vascular 437             55.13             2022 43    
3 Angiología y Cirugía Vascular 429             53.99             2022 42    
4 Angiología y Cirugía Vascular 419             56.29             2022 41    
5 Angiología y Cirugía Vascular 418             55.82             2022 40    
  CODCNH
1 280148
2 280148
3 280148
4 280148
5 280148

2. Catálogo de hospitales incluyendo área

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [9]:
# file_data_02 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [10]:
iFile_02 <- "CU_25_05_05_01_hospitales.csv"
file_data_02 <- paste0(iPath, iFile_02)

if(file.exists(file_data_02)){
    cat("Se leerán datos del archivo: ", file_data_02)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leerán datos del archivo:  Data/Input/CU_25_05_05_01_hospitales.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [11]:
data_02 <- read_csv(file_data_02)


Rows: 87 Columns: 24
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (6): Hospital, id_area, nombre_area, Municipio, Clase, Dependencia
dbl (18): CODCNH, cmunicipio, CAMAS, TAC, RM, GAM, HEM, ASD, LIT, BCO, ALI, ...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [12]:
data_02 |> glimpse()

Rows: 87
Columns: 24
$ CODCNH      <dbl> 281371, 280745, 281124, 281071, 281292, 281281, 281090, 28…
$ Hospital    <chr> "HOSPITAL HM VALLES", "HOSPITAL UNIVERSITARIO PRINCIPE DE …
$ id_area     <chr> "03", "03", "08", "08", "11", "01", "06", "06", "06", "11"…
$ nombre_area <chr> "Este", "Este", "Sur-Oeste I", "Sur-Oeste I", "Sur Ii", "S…
$ cmunicipio  <dbl> 280053, 280053, 280072, 280072, 280133, 280148, 280225, 28…
$ Municipio   <chr> "Alcalá de Henares", "Alcalá de Henares", "Alcorcón", "Alc…
$ CAMAS       <dbl> 20, 507, 81, 402, 98, 132, 170, 53, 192, 339, 344, 217, 13…
$ Clase       <chr> "Otros Centros con Internamiento", "Hospitales Generales",…
$ Dependencia <chr> "PRIVADOS", "SERVICIOS E INSTITUTOS DE SALUD DE LAS COMUNI…
$ TAC         <dbl> 1, 3, 1, 2, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 2, 0, 0, 0, 0…
$ RM          <dbl> 1, 2, 2, 2, 1, 1, 3, 2, 0, 0, 0, 2, 2, 1, 2, 2, 0, 0, 0, 0…
$ GAM         <dbl> 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0…
$ HEM         <dbl>

Muestra de datos:

In [13]:
data_02 |> slice_head(n = 5)

CODCNH Hospital                                    id_area nombre_area
1 281371 HOSPITAL HM VALLES                          03      Este       
2 280745 HOSPITAL UNIVERSITARIO PRINCIPE DE ASTURIAS 03      Este       
3 281124 HOSPITAL QUIRONSALUD SUR                    08      Sur-Oeste I
4 281071 HOSPITAL UNIVERSITARIO FUNDACION ALCORCON   08      Sur-Oeste I
5 281292 HOSPITAL DEL TAJO S.A.                      11      Sur Ii     
  cmunicipio Municipio         CAMAS Clase                          
1 280053     Alcalá de Henares  20   Otros Centros con Internamiento
2 280053     Alcalá de Henares 507   Hospitales Generales           
3 280072     Alcorcón           81   Hospitales Generales           
4 280072     Alcorcón          402   Hospitales Generales           
5 280133     Aranjuez           98   Hospitales Generales           
  Dependencia                                                  TAC ... LIT BCO
1 PRIVADOS                                                     1   ... 0   0  
2 SERVICIOS E INSTITUTOS DE SALUD DE LAS COMUNIDADES AUTÓNOMAS 3   ... 0   0  
3 PRIVADOS                                                     1   ... 0   0  
4 SERVICIOS E INSTITUTOS DE SALUD DE LAS COMUNIDADES AUTÓNOMAS 2   ... 1   0  
5 SERVICIOS E INSTITUTOS DE SALUD DE LAS COMUNIDADES AUTÓNOMAS 1   ... 0   0  
  ALI SPECT PET MAMOS DO DIAL X         Y       
1 0   0     0   1     1   0   -3.367372 40.48302
2 0   0     0   2     1  28   -3.347602 40.51025
3 0   0     0   1     1   0   -3.849035 40.34248
4 0   0     0   2     1  28   -3.837394 40.34887
5 0   0     0   1     0  13   -3.610795 40.05726

3. Indicadores áreas

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [14]:
# file_data_03 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [15]:
iFile_03 <- "CU_25_05_06_indicadores_area.csv"
file_data_03 <- paste0(iPath, iFile_03)

if(file.exists(file_data_03)){
    cat("Se leerán datos del archivo: ", file_data_03)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leerán datos del archivo:  Data/Input/CU_25_05_06_indicadores_area.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [16]:
data_03 <- read_csv(file_data_03)


Rows: 11 Columns: 17
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (2): id_area, nombre_area
dbl (15): t3_1, t1_1, t2_1, t2_2, t4_1, t4_2, t4_3, t5_1, t6_1, t7_1, t8_1, ...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [17]:
data_03 |> glimpse()

Rows: 11
Columns: 17
$ id_area     <chr> "01", "02", "03", "04", "05", "06", "07", "08", "09", "10"…
$ nombre_area <chr> "Sur-Este", "Centro-Norte", "Este", "Noreste", "Norte", "O…
$ t3_1        <dbl> 41.89330, 44.85910, 41.16384, 43.50412, 41.38840, 40.89333…
$ t1_1        <dbl> 862340, 404085, 383079, 615157, 912485, 763506, 511605, 53…
$ t2_1        <dbl> 0.5207432, 0.5419660, 0.5091281, 0.5321335, 0.5202002, 0.5…
$ t2_2        <dbl> 0.4792568, 0.4580340, 0.4908719, 0.4678665, 0.4791324, 0.4…
$ t4_1        <dbl> 0.1585425, 0.1293202, 0.1671827, 0.1495244, 0.1700457, 0.1…
$ t4_2        <dbl> 0.6732152, 0.6456867, 0.6697046, 0.6530478, 0.6590844, 0.6…
$ t4_3        <dbl> 0.1682410, 0.2249913, 0.1631133, 0.1974297, 0.1702033, 0.1…
$ t5_1        <dbl> 0.1399664, 0.1279791, 0.1520691, 0.1266269, 0.1248427, 0.1…
$ t6_1        <dbl> 0.2042778, 0.1854376, 0.1903965, 0.2132414, 0.1950003, 0.1…
$ t7_1        <dbl> 0.05120722, 0.05930819, 0.05762767, 0.05909755, 0.06406634…
$ t8_1        <dbl>

Muestra de datos:

In [18]:
data_03 |> slice_head(n = 5)

id_area nombre_area  t3_1     t1_1   t2_1      t2_2      t4_1      t4_2     
1 01      Sur-Este     41.89330 862340 0.5207432 0.4792568 0.1585425 0.6732152
2 02      Centro-Norte 44.85910 404085 0.5419660 0.4580340 0.1293202 0.6456867
3 03      Este         41.16384 383079 0.5091281 0.4908719 0.1671827 0.6697046
4 04      Noreste      43.50412 615157 0.5321335 0.4678665 0.1495244 0.6530478
5 05      Norte        41.38840 912485 0.5202002 0.4791324 0.1700457 0.6590844
  t4_3      t5_1      t6_1      t7_1       t8_1       t9_1      t10_1    
1 0.1682410 0.1399664 0.2042778 0.05120722 0.03895576 0.3600324 0.1440813
2 0.2249913 0.1279791 0.1854376 0.05930819 0.05200647 0.5572380 0.0948611
3 0.1631133 0.1520691 0.1903965 0.05762767 0.04557209 0.3196460 0.1452069
4 0.1974297 0.1266269 0.2132414 0.05909755 0.04845935 0.4600102 0.1154783
5 0.1702033 0.1248427 0.1950003 0.06406634 0.05318346 0.4872258 0.1032815
  t11_1     t12_1    
1 0.5193163 0.6047756
2 0.5069128 0.5605799
3 0.5250289 0.6109222
4 0.5117620 0.5773072
5 0.5420084 0.6035461

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

Se han importado en el apartado Data Load anterior:

* Listas de espera reales
* Catálogo de hospitales

Incluir apartados si procede para: Extracción de datos (select, filter), Transformación de datos, (mutate, joins, ...). Si es necesario tratar datos perdidos, indicarlo también en NB 09.2

Si no aplica: Estos datos no requieren tareas de este tipo.

#### Data transformation

* Unión de tablas y agrupar por área.


In [19]:
tdata_01 <- data_01 |>
    inner_join(data_02 |> select(CODCNH, id_area, nombre_area), by = "CODCNH") |>
    group_by(id_area, nombre_area, Hospital, Especialidad) |> 
    summarise(total_pacientes = mean(total_pacientes, na.rm = TRUE),
            media_tiempo_dias = mean(media_tiempo_dias, na.rm = TRUE)) |>
    ungroup()

`summarise()` has grouped output by 'id_area', 'nombre_area', 'Hospital'. You
can override using the `.groups` argument.


In [20]:
tdata_01 |> glimpse()


Rows: 464
Columns: 6
$ id_area           <chr> "01", "01", "01", "01", "01", "01", "01", "01", "01"…
$ nombre_area       <chr> "Sur-Este", "Sur-Este", "Sur-Este", "Sur-Este", "Sur…
$ Hospital          <chr> "HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS", "HOSPI…
$ Especialidad      <chr> "Angiología y Cirugía Vascular", "Cirugía Cardiaca",…
$ total_pacientes   <dbl> 0.00000, 0.00000, 0.00000, 41.93277, 315.46218, 93.2…
$ media_tiempo_dias <dbl> 0.00000, 0.00000, 0.00000, 39.46882, 53.02748, 60.17…


In [21]:
tdata_01 |> slice_head(n = 5)

id_area nombre_area Hospital                                  
1 01      Sur-Este    HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS
2 01      Sur-Este    HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS
3 01      Sur-Este    HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS
4 01      Sur-Este    HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS
5 01      Sur-Este    HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS
  Especialidad                            total_pacientes media_tiempo_dias
1 Angiología y Cirugía Vascular             0.00000        0.00000         
2 Cirugía Cardiaca                          0.00000        0.00000         
3 Cirugía General y del Aparato Digestivo   0.00000        0.00000         
4 Cirugía Oral y Maxilofacial              41.93277       39.46882         
5 Cirugía Pediátrica General              315.46218       53.02748

## Synthetic Data Generation

* Asignar capacidad
* Para cada hospital y especialidad, asignar la capacidad. Se asume la misma capacidad en el tiempo, personas / tiempo

In [22]:
sdata_01 <- tdata_01 |>
    mutate(capacidad = ceiling(total_pacientes/media_tiempo_dias)) |> 
    replace_na(list(capacidad = 0)) 

In [23]:
sdata_01 |> glimpse()

Rows: 464
Columns: 7
$ id_area           <chr> "01", "01", "01", "01", "01", "01", "01", "01", "01"…
$ nombre_area       <chr> "Sur-Este", "Sur-Este", "Sur-Este", "Sur-Este", "Sur…
$ Hospital          <chr> "HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS", "HOSPI…
$ Especialidad      <chr> "Angiología y Cirugía Vascular", "Cirugía Cardiaca",…
$ total_pacientes   <dbl> 0.00000, 0.00000, 0.00000, 41.93277, 315.46218, 93.2…
$ media_tiempo_dias <dbl> 0.00000, 0.00000, 0.00000, 39.46882, 53.02748, 60.17…
$ capacidad         <dbl> 0, 0, 0, 2, 6, 2, 0, 0, 0, 1, 4, 2, 3, 23, 5, 3, 0, …


In [24]:
sdata_01 |> slice_head(n = 5)

id_area nombre_area Hospital                                  
1 01      Sur-Este    HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS
2 01      Sur-Este    HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS
3 01      Sur-Este    HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS
4 01      Sur-Este    HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS
5 01      Sur-Este    HOSPITAL INFANTIL UNIVERSITARIO NIÑO JESUS
  Especialidad                            total_pacientes media_tiempo_dias
1 Angiología y Cirugía Vascular             0.00000        0.00000         
2 Cirugía Cardiaca                          0.00000        0.00000         
3 Cirugía General y del Aparato Digestivo   0.00000        0.00000         
4 Cirugía Oral y Maxilofacial              41.93277       39.46882         
5 Cirugía Pediátrica General              315.46218       53.02748         
  capacidad
1 0        
2 0        
3 0        
4 2        
5 6

### Data transformation

* Agrupar capacidad por área

In [25]:
tdata_01_area <- sdata_01 |> 
  group_by(id_area, nombre_area, Especialidad) |> 
  summarise(capacidad = sum(capacidad)) |>
  ungroup()

`summarise()` has grouped output by 'id_area', 'nombre_area'. You can override
using the `.groups` argument.


In [26]:
tdata_01_area |> glimpse()

Rows: 160
Columns: 4
$ id_area      <chr> "01", "01", "01", "01", "01", "01", "01", "01", "01", "01…
$ nombre_area  <chr> "Sur-Este", "Sur-Este", "Sur-Este", "Sur-Este", "Sur-Este…
$ Especialidad <chr> "Angiología y Cirugía Vascular", "Cirugía Cardiaca", "Cir…
$ capacidad    <dbl> 0, 0, 20, 2, 6, 2, 0, 9, 6, 1, 4, 27, 11, 99, 18, 14, 6, …


In [27]:
tdata_01_area |> slice_head(n = 5)

id_area nombre_area Especialidad                            capacidad
1 01      Sur-Este    Angiología y Cirugía Vascular            0       
2 01      Sur-Este    Cirugía Cardiaca                         0       
3 01      Sur-Este    Cirugía General y del Aparato Digestivo 20       
4 01      Sur-Este    Cirugía Oral y Maxilofacial              2       
5 01      Sur-Este    Cirugía Pediátrica General               6

## Fake Data Generation

Proceso de simulación de pacientes, consultas y hospitalizaciones:

1. Unión de lista con indicadores
2. Simulación de pacientes teniendo en cuenta tamaño de la cola y población del área
3. Simulación de consultas teniendo en cuenta el tamaño de la cola
4. Simulación de hospitalizaciones teneiendo en cuenta el tamaño de la cola
5. Quitar información de indicadores

In [28]:
set.seed(1)
tdata_03 <- data_01 |> 
       inner_join(data_02 |> select(CODCNH, id_area, nombre_area), by = c("CODCNH")) |>
  inner_join(data_03, by = c("id_area", "nombre_area")) |> 
  mutate(pacientes = ceiling((max(c(total_pacientes, 10), na.rm = TRUE)/t1_1)*(1+rnorm(n(), 0.6, 0.1))*t1_1*0.1),
         consultas = ceiling(total_pacientes*1.5 + sample(c(-1, 1), 1)*ceiling(rnorm(n(), 50, 10))),
         hospitalizaciones = ceiling(total_pacientes*runif(n(), 0, 0.05))) |>
  select(-(t3_1:t12_1))

In [29]:
tdata_03 |> glimpse()

Rows: 55,680
Columns: 12
$ Hospital          <chr> "HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA A…
$ Especialidad      <chr> "Angiología y Cirugía Vascular", "Angiología y Cirug…
$ total_pacientes   <dbl> 439, 437, 429, 419, 418, NA, 408, 399, 396, 393, 378…
$ media_tiempo_dias <dbl> 56.29, 55.13, 53.99, 56.29, 55.82, NA, 57.45, 57.61,…
$ ano               <dbl> 2022, 2022, 2022, 2022, 2022, 2022, 2022, 2022, 2022…
$ semana            <dbl> 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, …
$ CODCNH            <dbl> 280148, 280148, 280148, 280148, 280148, 280148, 2801…
$ id_area           <chr> "05", "05", "05", "05", "05", "05", "05", "05", "05"…
$ nombre_area       <chr> "Norte", "Norte", "Norte", "Norte", "Norte", "Norte"…
$ pacientes         <dbl> 1392, 1465, 1373, 1593, 1478, 1374, 1493, 1515, 1501…
$ consultas         <dbl> 717, 703, 684, 686, 682, NA, 656, 657, 646, 641, 625…
$ hospitalizaciones <dbl> 8, 3, 12, 3, 17, NA, 10, 20, 14, 8, 19, 8, 18, 6, 5,…


### Data Extract

* Consolidar todos los datos en una sola tabla

In [30]:
tdata_03 |> select(CODCNH, pacientes:hospitalizaciones, ano, semana)

CODCNH pacientes consultas hospitalizaciones ano  semana
1     280148 1392      717        8                2022 44    
2     280148 1465      703        3                2022 43    
3     280148 1373      684       12                2022 42    
4     280148 1593      686        3                2022 41    
5     280148 1478      682       17                2022 40    
6     280148 1374       NA       NA                2022 39    
7     280148 1493      656       10                2022 38    
8     280148 1515      657       20                2022 37    
9     280148 1501      646       14                2022 36    
10    280148 1421      641        8                2022 35    
11    280148 1585      625       19                2022 34    
12    280148 1484      619        8                2022 33    
13    280148 1392      607       18                2022 32    
14    280148 1248      581        6                2022 31    
15    280148 1550      553        5                2022 30    
16    280148 1445      539        8                2022 29    
17    280148 1447      496       15                2022 28    
18    280148 1534      495       14                2022 27    
19    280148 1523      484        4                2022 25    
20    280148 1502      478        5                2022 24    
21    280148 1532      460        6                2022 23    
22    280148 1519      478        6                2022 22    
23    280148 1455      514        6                2022 21    
24    280148 1269      569       14                2022 20    
25    280148 1505      572       15                2022 19    
26    280148 1444      578        6                2022 18    
27    280148 1435      598        4                2022 17    
28    280148 1316      624        1                2022 16    
29    280148 1405      630       15                2022 15    
30    280148 1486      625       10                2022 14    
...   ...    ...       ...       ...               ...  ...   
55651 280838 1275      3742       62               2021  1    
55652 280838 1445      4156       86               2020 51    
55653 280838 1499      4128      104               2020 50    
55654 280838 1396      4017        8               2020 49    
55655 280838 1380      3932      109               2020 48    
55656 280838 1566      3884       71               2020 47    
55657 280838 1506      3855       56               2020 46    
55658 280838 1317      3740      110               2020 45    
55659 280838 1491      3729       81               2020 43    
55660 280838 1439      3660       62               2020 42    
55661 280838 1369      3588       23               2020 41    
55662 280838 1556      3438       65               2020 40    
55663 280838 1471      3331       64               2020 39    
55664 280838 1488      3254       70               2020 38    
55665 280838 1388      3225      100               2020 37    
55666 280838 1434      3187       30               2020 36    
55667 280838 1498      3263       22               2020 35    
55668 280838 1620      3258       59               2020 34    
55669 280838 1489      3254       33               2020 33    
55670 280838 1408      3252       66               2020 32    
55671 280838 1455      3200       71               2020 31    
55672 280838 1318      3172       21               2020 30    
55673 280838 1570      3315       27               2020 10    
55674 280838 1488      3230       38               2020  9    
55675 280838 1505      3250       36               2020  7    
55676 280838 1424      3175       98               2020  6    
55677 280838 1462      3187       86               2020  5    
55678 280838 1461      3466       83               2020  4    
55679 280838 1436      3466        3               2020  3    
55680 280838 1352      3203       30               2020  2

In [31]:
a <- data_01 |> 
  inner_join(data_02 |> select(-Hospital), by = c("CODCNH")) |>
  inner_join(data_03, by = c("id_area", "nombre_area")) |>
  inner_join(tdata_01_area,  by = c("id_area", "nombre_area", "Especialidad"))
 b <-  tdata_03 |> select(CODCNH, pacientes:hospitalizaciones, ano, semana, Especialidad)
 names(a); nrow(a)
 names(b); nrow(b)
 c <- inner_join(a, b)
 names(c); nrow(c)
 nrow(c)/nrow(b)
 length(unique(a$Especialidad))

[1] "Hospital"          "Especialidad"      "total_pacientes"  
 [4] "media_tiempo_dias" "ano"               "semana"           
 [7] "CODCNH"            "id_area"           "nombre_area"      
[10] "cmunicipio"        "Municipio"         "CAMAS"            
[13] "Clase"             "Dependencia"       "TAC"              
[16] "RM"                "GAM"               "HEM"              
[19] "ASD"               "LIT"               "BCO"              
[22] "ALI"               "SPECT"             "PET"              
[25] "MAMOS"             "DO"                "DIAL"             
[28] "X"                 "Y"                 "t3_1"             
[31] "t1_1"              "t2_1"              "t2_2"             
[34] "t4_1"              "t4_2"              "t4_3"             
[37] "t5_1"              "t6_1"              "t7_1"             
[40] "t8_1"              "t9_1"              "t10_1"            
[43] "t11_1"             "t12_1"             "capacidad"

[1] 55680

[1] "CODCNH"            "pacientes"         "consultas"        
[4] "hospitalizaciones" "ano"               "semana"           
[7] "Especialidad"

[1] 55680

Joining, by = c("Especialidad", "ano", "semana", "CODCNH")


[1] "Hospital"          "Especialidad"      "total_pacientes"  
 [4] "media_tiempo_dias" "ano"               "semana"           
 [7] "CODCNH"            "id_area"           "nombre_area"      
[10] "cmunicipio"        "Municipio"         "CAMAS"            
[13] "Clase"             "Dependencia"       "TAC"              
[16] "RM"                "GAM"               "HEM"              
[19] "ASD"               "LIT"               "BCO"              
[22] "ALI"               "SPECT"             "PET"              
[25] "MAMOS"             "DO"                "DIAL"             
[28] "X"                 "Y"                 "t3_1"             
[31] "t1_1"              "t2_1"              "t2_2"             
[34] "t4_1"              "t4_2"              "t4_3"             
[37] "t5_1"              "t6_1"              "t7_1"             
[40] "t8_1"              "t9_1"              "t10_1"            
[43] "t11_1"             "t12_1"             "capacidad"        
[46] "pacientes"         "consultas"         "hospitalizaciones"

[1] 55680

[1] 1

[1] 16

In [32]:
data_completo <- data_01 |> 
  inner_join(data_02 |> select(-Hospital), by = c("CODCNH")) |>
  inner_join(data_03, by = c("id_area", "nombre_area")) |>
  inner_join(tdata_01_area,  by = c("id_area", "nombre_area", "Especialidad")) |> 
  inner_join(tdata_03 |> select(CODCNH, Especialidad, ano, semana, pacientes:hospitalizaciones),
    by = c("CODCNH", "Especialidad", "ano", "semana"))
  

In [33]:

data_completo |> glimpse()

Rows: 55,680
Columns: 48
$ Hospital          <chr> "HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA A…
$ Especialidad      <chr> "Angiología y Cirugía Vascular", "Angiología y Cirug…
$ total_pacientes   <dbl> 439, 437, 429, 419, 418, NA, 408, 399, 396, 393, 378…
$ media_tiempo_dias <dbl> 56.29, 55.13, 53.99, 56.29, 55.82, NA, 57.45, 57.61,…
$ ano               <dbl> 2022, 2022, 2022, 2022, 2022, 2022, 2022, 2022, 2022…
$ semana            <dbl> 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, …
$ CODCNH            <dbl> 280148, 280148, 280148, 280148, 280148, 280148, 2801…
$ id_area           <chr> "05", "05", "05", "05", "05", "05", "05", "05", "05"…
$ nombre_area       <chr> "Norte", "Norte", "Norte", "Norte", "Norte", "Norte"…
$ cmunicipio        <dbl> 280796, 280796, 280796, 280796, 280796, 280796, 2807…
$ Municipio         <chr> "Madrid", "Madrid", "Madrid", "Madrid", "Madrid", "M…
$ CAMAS             <dbl> 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 15…
$ Clase        

In [34]:
data_completo |> slice_head(n = 5)

Hospital                                               
1 HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA ADELA
2 HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA ADELA
3 HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA ADELA
4 HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA ADELA
5 HOSPITAL CENTRAL DE LA CRUZ ROJA SAN JOSE Y SANTA ADELA
  Especialidad                  total_pacientes media_tiempo_dias ano  semana
1 Angiología y Cirugía Vascular 439             56.29             2022 44    
2 Angiología y Cirugía Vascular 437             55.13             2022 43    
3 Angiología y Cirugía Vascular 429             53.99             2022 42    
4 Angiología y Cirugía Vascular 419             56.29             2022 41    
5 Angiología y Cirugía Vascular 418             55.82             2022 40    
  CODCNH id_area nombre_area cmunicipio ... t7_1       t8_1       t9_1     
1 280148 05      Norte       280796     ... 0.06406634 0.05318346 0.4872258
2 280148 05      Norte       280796     ... 0.06406634 0.05318346 0.4872258
3 280148 05      Norte       280796     ... 0.06406634 0.05318346 0.4872258
4 280148 05      Norte       280796     ... 0.06406634 0.05318346 0.4872258
5 280148 05      Norte       280796     ... 0.06406634 0.05318346 0.4872258
  t10_1     t11_1     t12_1     capacidad pacientes consultas hospitalizaciones
1 0.1032815 0.5420084 0.6035461 14        1392      717        8               
2 0.1032815 0.5420084 0.6035461 14        1465      703        3               
3 0.1032815 0.5420084 0.6035461 14        1373      684       12               
4 0.1032815 0.5420084 0.6035461 14        1593      686        3               
5 0.1032815 0.5420084 0.6035461 14        1478      682       17

## Open Data

No aplica

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

1. Datos de capacidad (solo por área y especialidad)

<font color='tomato'> Identificamos los datos a guardar</font>

In [35]:
data_to_save_01 <- tdata_01_area


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [36]:
caso <- "CU_25"
proceso <- '_05'
tarea <- "_07"
archivo <- "_01"
proper <- "_capacidad"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [37]:
# file_save_01 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_01 <- paste0(oPath, file_save_01)
# write_csv(data_to_save_01, path_out_01)

# cat('File saved as: ')
# path_out_02

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [38]:
file_save_01 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_01 <- paste0(oPath, file_save_01)
write_csv(data_to_save_01, path_out_01)

cat('File saved as: ')
path_out_01

File saved as: 

[1] "Data/Output/CU_25_05_07_01_capacidad.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [39]:
path_in_01 <- paste0(iPath, file_save_01)
file.copy(path_out_01, path_in_01, overwrite = TRUE)

[1] TRUE

2. Listas de espera con datos sanitarios

<font color='tomato'> Identificamos los datos a guardar</font>

In [40]:
data_to_save_02 <- tdata_03


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [41]:
archivo <- "_02"
proper <- "_lista_espera"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [42]:
# file_save_02 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_02 <- paste0(oPath, file_save_02)
# write_csv(data_to_save_02, path_out_02)

# cat('File saved as: ')
# path_out_02

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [43]:
file_save_02 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_02 <- paste0(oPath, file_save_02)
write_csv(data_to_save_02, path_out_02)

cat('File saved as: ')
path_out_02

File saved as: 

[1] "Data/Output/CU_25_05_07_02_lista_espera.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [44]:
path_in_02 <- paste0(iPath, file_save_02)
file.copy(path_out_02, path_in_02, overwrite = TRUE)

[1] TRUE

3. Datos completos en un archivo

<font color='tomato'> Identificamos los datos a guardar</font>

In [45]:
data_to_save_03 <- data_completo


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [46]:
archivo <- "_03"
proper <- "_lista_espera_completo"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [47]:
# file_save_03 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_03 <- paste0(oPath, file_save_03)
# write_csv(data_to_save_03, path_out_03)

# cat('File saved as: ')
# path_out_03

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [48]:
file_save_03 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_03 <- paste0(oPath, file_save_03)
write_csv(data_to_save_03, path_out_03)

cat('File saved as: ')
path_out_03

File saved as: 

[1] "Data/Output/CU_25_05_07_03_lista_espera_completo.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [49]:
path_in_03 <- paste0(iPath, file_save_03)
file.copy(path_out_03, path_in_03, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

Para que funcione este código se necesita:

* Las rutas de archivos `Data/Input` y `Data/Output` deben existir (relativas a la ruta del _notebook_)
* El paquete tcltk instalado para seleccionar archivos interactivamente. No se necesita en producción.
* Los paquetes readr, dplyr, tidyr deben estar instalados.


### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* readr 2.1.3
* dplyr 1.0.10
* tidyr 1.3.0

### Data structures

#### Objeto `tdata_01_area`

- Hay 160 filas con información de las siguientes variables:
	* id_area
	* nombre_area
	* Especialidad
	* capacidad


#### Observaciones generales sobre los datos

* Se dispone de la capacidad global para cada especialidad en cada área
* Se entiende que la capacidad dependerá de los quirófanos/equipos de especialistas disponibles en los hospitales


#### Objeto `tdata_03`

- Hay 464 filas con información de las siguientes variables:
	* id_area
	* nombre_area
	* Hospital
	* Especialidad
	* total_pacientes
	* media_tiempo_dias
	* capacidad
	* pacientes
	* consultas
	* hospitalizaciones


In [50]:
cat("- Hay", nrow(data_to_save_02), "filas con información de las siguientes variables:\n\t*", paste(colnames(data_to_save_02), collapse = "\n\t* "))

- Hay 55680 filas con información de las siguientes variables:
	* Hospital
	* Especialidad
	* total_pacientes
	* media_tiempo_dias
	* ano
	* semana
	* CODCNH
	* id_area
	* nombre_area
	* pacientes
	* consultas
	* hospitalizaciones

#### Objeto `data_completo`

- Los datos con distinta granularidad se repiten
- Hay 55680 filas con información de las siguientes variables:
	* Hospital
	* Especialidad
	* total_pacientes
	* media_tiempo_dias
	* ano
	* semana
	* CODCNH
	* id_area
	* nombre_area
	* cmunicipio
	* Municipio
	* CAMAS
	* Clase
	* Dependencia
	* TAC
	* RM
	* GAM
	* HEM
	* ASD
	* LIT
	* BCO
	* ALI
	* SPECT
	* PET
	* MAMOS
	* DO
	* DIAL
	* X
	* Y
	* t3_1
	* t1_1
	* t2_1
	* t2_2
	* t4_1
	* t4_2
	* t4_3
	* t5_1
	* t6_1
	* t7_1
	* t8_1
	* t9_1
	* t10_1
	* t11_1
	* t12_1
	* capacidad
	* pacientes
	* consultas
	* hospitalizaciones


#### Observaciones generales sobre los datos

* Se dispone de la capacidad global para cada especialidad en cada área
* Se entiende que la capacidad dependerá de los quirófanos/equipos de especialistas disponibles en los hospitales


### Consideraciones para despliegue en piloto

* Los datos con distinta resolución se guardan en archivos distintos. Se han unido también los datos en una tabla completa repitiendo datos en filas cuando procede. A la hora de extraer estos datos hay que estar seguro de la operación de agregación a usar (media, suma, otra)

### Consideraciones para despliegue en producción

* Se deben crear los procesos ETL en producción necesarios para que los datos de entrada estén actualizados
* Se ha asumido el área sanitaria como agregación de hospitales en todo el caso. En el despliegue en producción se podría cambiar modificando los notebooks de la forma adecuada

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se han asignado capacidades por especialidad y zona
- Se han simulado datos sanitarios por zona y semana

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se deben unir los datos a la tabla completa

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [52]:
# incluir código